### PR TITLE
add imu module

### LINKED ISSUE #53: integrate bno055 library into imu module
the imu module currently just sends a single i2c command, we need to create an instance of the bno055 library and pass in i2c functions.

### PR DESCRIPTION
**Pull request by harneetMac**
Closes #53

### CODE REVIEW

File: firmware/applications/SignBuddy/imu.c [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+#include "i2c.h"
+#include "logger.h"
+#include "bno055.h"
+
+#define PROCESS_PERIOD_MS    1000
+
+struct bno055_t bno055;
+struct bno055_accel_t bno055_accel_xyz;
+struct bno055_mag_t bno055_mag_xyz;
+struct bno055_gyro_t bno055_gyro_xyz;
```
Comment (adambujak) [inline firmware/applications/SignBuddy/imu.c:14]: add these to the state struct

File: firmware/applications/SignBuddy/imu.c [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  if (system_time_cmp_ms(s.last_ticks, time) < PROCESS_PERIOD_MS) {
+    return;
+  }
+  s.last_ticks = time;
+
+  uint8_t read_val;
+  i2c_read(&s.i2c_instance, 0x28 << 1, 0x0, &read_val, 1);
+
+  LOG_INFO("i2c read: %d\r\n", read_val);
+  LOG_DEBUG("imu process\r\n");
```
Comment (adambujak) [inline firmware/applications/SignBuddy/imu.c:124]: we can get rid of this now

File: firmware/applications/SignBuddy/main.c [hunk 3]
```diff
@@ -52,8 +53,10 @@ void sysclk_init(void)
   LL_Init1msTick(SYSCLK_FREQ);
 
   LL_SetSystemCoreClock(SYSCLK_FREQ);
+
```
Comment (adambujak) [inline firmware/applications/SignBuddy/main.c:56]: remove this extra line

File: firmware/applications/SignBuddy/nucleo_board.h [hunk 4]
```diff
@@ -69,4 +70,12 @@
...
 #define FLEX_ADC_GPIO_CLK_EN()    LL_IOP_GRP1_EnableClock(LL_IOP_GRP1_PERIPH_GPIOA)
 
+/* I2C */
+#define IMU_I2C                   I2C1
+#define IMU_I2C_CLK_SRC()         LL_RCC_SetI2CClockSource(LL_RCC_I2C1_CLKSOURCE_PCLK1);
+
+/* I2C */
+#define IMU_I2C                   I2C1
+#define IMU_I2C_CLK_SRC()         LL_RCC_SetI2CClockSource(LL_RCC_I2C1_CLKSOURCE_PCLK1);
+
```
Comment (adambujak) [inline firmware/applications/SignBuddy/nucleo_board.h:80]: here twice

File: firmware/applications/SignBuddy/imu.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  i2c_config.DigitalFilter = 0;
+  i2c_config.OwnAddress1 = 0;
+  i2c_config.TypeAcknowledge = LL_I2C_ACK;
+  i2c_config.OwnAddrSize = LL_I2C_OWNADDRESS1_7BIT;
+  LL_I2C_SetOwnAddress2(IMU_I2C, 0, LL_I2C_OWNADDRESS2_NOMASK);
+
+  i2c_init(&s.i2c_instance, IMU_I2C, &i2c_config);
+}
+
+static inline int8_t write(uint8_t dev_addr, uint8_t reg_addr, uint8_t *reg_data, uint8_t cnt)
```
Comment (adambujak) [inline firmware/applications/SignBuddy/imu.c:54]: match these parameter names to the i2c module's naming

File: firmware/applications/SignBuddy/imu.c [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+typedef struct {
+  uint32_t                last_ticks;
+  i2c_t                   i2c_instance;
+  struct   bno055_t       bno055;
+  struct   bno055_accel_t bno055_accel_xyz;
+  struct   bno055_mag_t   bno055_mag_xyz;
+  struct   bno055_gyro_t  bno055_gyro_xyz;
+} state_t;
+
+static state_t s;
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:20]: Rename to state

File: firmware/applications/SignBuddy/imu.c [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  IMU_I2C_GPIO_CLK_EN();
+
+  LL_GPIO_InitTypeDef gpio_config = { 0 };
+  gpio_config.Pin = LL_GPIO_PIN_6 | LL_GPIO_PIN_7;
+  gpio_config.Mode = LL_GPIO_MODE_ALTERNATE;
+  gpio_config.Speed = LL_GPIO_SPEED_FREQ_VERY_HIGH;
+  gpio_config.OutputType = LL_GPIO_OUTPUT_OPENDRAIN;
+  gpio_config.Pull = LL_GPIO_PULL_NO;
+  gpio_config.Alternate = LL_GPIO_AF_1;
+  LL_GPIO_Init(GPIOB, &gpio_config);
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:34]: Abstract pins and alternate function to board header

File: firmware/applications/SignBuddy/imu.c [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+  LL_I2C_InitTypeDef i2c_config = { 0 };
+  LL_I2C_EnableAutoEndMode(IMU_I2C);
+  LL_I2C_DisableOwnAddress2(IMU_I2C);
+  LL_I2C_DisableGeneralCall(IMU_I2C);
+  LL_I2C_EnableClockStretching(IMU_I2C);
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/imu.c:40]: Do we need to call these functions here and in i2c_init?

File: firmware/applications/SignBuddy/imu.c (hunk 6 above)
Comment (harneetMac) [reply firmware/applications/SignBuddy/imu.c:20]: you mean rename s to state (s -> state) or also the struct member names (bno055_gyro_xyz -> state_bno055_gyro_xyz)?

File: firmware/applications/SignBuddy/imu.c (hunk 7 above)
Comment (harneetMac) [reply firmware/applications/SignBuddy/imu.c:34]: sure, will do.

File: firmware/applications/SignBuddy/imu.c (hunk 6 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/imu.c:20]: I meant change `s` to `state`

File: firmware/applications/SignBuddy/imu.c (hunk 8 above)
Comment (adambujak) [reply firmware/applications/SignBuddy/imu.c:40]: no we don't. good catch. I think we can put that into the i2c init

File: firmware/applications/SignBuddy/imu.c [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+{
+  i2c_write(&state.i2c_instance, slave_addr << 1, reg_addr, data, (uint16_t) length);
+  return 0;
+}
+
+static inline int8_t read(uint8_t slave_addr, uint8_t reg_addr, uint8_t *data, uint8_t length)
+{
+  i2c_read(&state.i2c_instance, slave_addr << 1, reg_addr, data, (uint16_t) length);
+  return 0;
+}
```
Comment (adambujak) [inline firmware/applications/SignBuddy/imu.c:63]: i think we should rename these since read/write are both standard c functions already

File: firmware/applications/SignBuddy/imu.c [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+typedef struct {
+  uint32_t                last_ticks;
+  i2c_t                   i2c_instance;
+  struct   bno055_t       bno055;
+  struct   bno055_accel_t bno055_accel_xyz;
+  struct   bno055_mag_t   bno055_mag_xyz;
+  struct   bno055_gyro_t  bno055_gyro_xyz;
+} state_t;
+
+static state_t state;
```
Comment (adambujak) [inline firmware/applications/SignBuddy/imu.c:20]: why did you rename this to state?

File: firmware/applications/SignBuddy/nucleo_board.h [hunk 11]
```diff
@@ -69,4 +70,14 @@
 #define FLEX_ADC_CLK_EN()         LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_ADC1)
 #define FLEX_ADC_GPIO_CLK_EN()    LL_IOP_GRP1_EnableClock(LL_IOP_GRP1_PERIPH_GPIOA)
 
+/* I2C */
+#define IMU_I2C                   I2C1
+#define IMU_I2C_SCL_PIN           LL_GPIO_PIN_6
+#define IMU_I2C_SDA_PIN           LL_GPIO_PIN_7
+#define IMU_I2C_GPIO_AF           LL_GPIO_AF_1
+#define IMU_I2C_GPIO_PORT         GPIOB
```
Comment (adambujak) [inline firmware/applications/SignBuddy/nucleo_board.h:78]: thanks, I forgot to abstract the gpio settings

File: firmware/applications/SignBuddy/imu.c (hunk 9 above)
Comment (harneetMac) [reply firmware/applications/SignBuddy/imu.c:63]: okay sure

File: firmware/applications/SignBuddy/imu.c (hunk 10 above)
Comment (harneetMac) [reply firmware/applications/SignBuddy/imu.c:20]: i thought it'd be a bit more descriptive.

Comment (stephen99scott) [reply firmware/applications/SignBuddy/imu.c:20]: That was my suggestion, I thought it was state in the sensors module too but looks like it's just s

Comment (adambujak) [reply firmware/applications/SignBuddy/imu.c:20]: it is more descriptive, but it's really just supposed to be a short prefix to encapsulate module variables

Comment (adambujak) [reply firmware/applications/SignBuddy/imu.c:20]: it was state in sensors once, but s is better I find

Comment (stephen99scott) [reply firmware/applications/SignBuddy/imu.c:20]: Okay let's change it back to s then to be consistent

**Review by stephen99scott — APPROVED** [review]
